enhancements/olm/opm-mirroring.md

```diff
@@ -0,0 +1,183 @@
+# Disconnected OLM
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+This enhancement outlines the changes necessary to `olm`, `operator-registry`, and `opm` to support taking operator catalogs and packaging them for installation in an offline cluster.
+
+This enhancement also covers how this process will change as our distribution mechanism changes and as appregistry support is deprecated.
+
+## Motivation
+
+Disconnected installation of OLM catalog data requires a different approach from other openshift components, because:
+
+ - The catalog of operator data currently exists in appregistry repositories (Quay.io)
+ - That catalog references a set of operators that are versioned out-of-sync with OpenShift
+ - Each version of each operator may require an additional set of `operand` images which need to be mirrored.
+
+The [current process]() for offlining a catalog involves many manual steps, and may fail to capture all images required for disconnected environments. 
+
+
+### Goals
+* Define a mechanism for operators to specify related operand images that should be collected for disconnected installation.
+* Define the changes required in `olm` and `operator-registry` to allow those images to be discovered
+* Define `opm` commands which can build and mirror all images required for mirroring to a disconnected envorinment
+
+### Non-Goals
+* Provide an api for fine-grained control over which operators / operands are mirrored. 
+
+## Proposal
+
+The basic approach will be to:
+
+- Allow operator manifests to specify [related images](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/contributors/design-proposals/related-images.md) in their manifests
+- Generate catalog container images that can replace the appregistry repositories
+- For each operator within the catalog image, mirror it to the disconnected environment
+
+
+### Specify related images
+
+We add a new field to the `ClusterServiceVersion` spec. 
+
+```yaml
+kind: ClusterServiceVersion 
+spec:
+  relatedImages:
+  - name: default
+    image: quay.io/coreos/etcd@sha256:12345 
+  - name: etcd-2.1.5
+    image: quay.io/coreos/etcd@sha256:12345 
+  - name: etcd-3.1.1
+    image: quay.io/coreos/etcd@sha256:12345  
+```
+
+This is covered in depth in the [related images](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/contributors/design-proposals/related-images.md) proposal for OLM.
+
+### Expose related images
+
+`operator-registry` has tools for building an index of operator manifests into a sqlite artifact, with interfaces provided for querying that database.
+
+A new table will be added to the sql schema:
+
+```sql
+	CREATE TABLE IF NOT EXISTS bundle_image (
```

Comment:
Maybe we want to call this `operator_images`? `bundle_image` is going to be confusing since we are also adding data to the table related to the operator bundle image, which is the container image that contains the manifests.

Reply:
Changed to `related_image`